Add sensors for current and potential to AVM FRITZ!SmartHome / Fritz!DECT

dreimer1986 opened the pull request:
Add sensors for current and potential to AVM FRITZ!SmartHome. Tested with Fritz!DECT 200

<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->

## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
It resolves my own feature request:
https://community.home-assistant.io/t/avm-fritz-smarthome-extract-not-just-watts-and-kwh-but-ampere-volt-too/411203

More precisely it adds a way to get the information about current and potential from a Fritz!DECT. These devices know these values as it's confirmed by the Web UI of the Fritz!Box and pyfritzhome itself. These few lines add both sensors. Tested locally and confirmed working on 5 Fritz!DECT.

This is my first pull request here EVER. I never did pull to more than a few small projects and I have no real python skills. This PR is the result of logic and reading documentation. So sorry if things are wrong somewhere ^^ 

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [X] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/22508

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [X] The code change is tested and works locally.
- [X] Local tests pass. **Your PR cannot be merged unless tests pass**
- [X] There is no commented out code in this PR.
- [ ] I have followed the [development checklist][dev-checklist]
- [ ] The code has been formatted using Black (`black --fast homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [X] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

dreimer1986 committed:
Update sensor.py

Add sensors for current and potential to AVM FRITZ!SmartHome. Tested with Fritz!DECT 200

homeassistant commented:
Hi @dreimer1986, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;core#69999).

Once you do that we will be able to review and accept this pull request.

Thanks!

dreimer1986 committed:
Update sensor.py

black, flake8 and isort checked

mib1185 commented on homeassistant/components/fritzbox/sensor.py:
```diff
@@ -87,6 +82,26 @@ class FritzSensorEntityDescription(
+        state_class=SensorStateClass.MEASUREMENT,
+        suitable=lambda device: device.has_powermeter,  # type: ignore[no-any-return]
+        native_value=lambda device: device.power / 1000 / device.voltage
+        if device.voltage
```
we need to check both properties to be available
```suggestion
        if device.power and device.voltage
```

mib1185 reviewed (changes requested):
Hi @dreimer1986 

many thanks for this contribution 👍 
The addition of these two sensors itself looks good (_besides one small comment below_), but please setup your development environment as described [here](https://developers.home-assistant.io/docs/development_environment) or at least make sure to follow our [style guidelines](https://developers.home-assistant.io/docs/development_guidelines)

mib1185 commented on homeassistant/components/fritzbox/sensor.py:
```diff
@@ -87,6 +82,26 @@ class FritzSensorEntityDescription(
+        device_class=SensorDeviceClass.VOLTAGE,
+        state_class=SensorStateClass.MEASUREMENT,
+        suitable=lambda device: device.has_powermeter,  # type: ignore[no-any-return]
+        native_value=lambda device: device.voltage if device.voltage else 0.0,
```
based on [API documention (_page 9_)](https://avm.de/fileadmin/user_upload/Global/Service/Schnittstellen/AHA-HTTP-Interface.pdf) `voltage` is returned as 0.001 volts, so needs to be divided by 1000 (_same like `power` property_)

mib1185 commented on homeassistant/components/fritzbox/sensor.py:
```diff
@@ -87,6 +82,26 @@ class FritzSensorEntityDescription(
+        device_class=SensorDeviceClass.CURRENT,
+        state_class=SensorStateClass.MEASUREMENT,
+        suitable=lambda device: device.has_powermeter,  # type: ignore[no-any-return]
+        native_value=lambda device: device.power / 1000 / device.voltage
```
since `power` and `voltage` are returned as 0.001, we do not need to divide by 1000

dreimer1986 committed:
Fix the change requests

Fixed the division by 1000 on two code parts to comply with the AVM API documention.
Added device.power to availability check.

mib1185 commented on homeassistant/components/fritzbox/sensor.py:
```diff
@@ -6,30 +6,25 @@
 from homeassistant.config_entries import ConfigEntry
-from homeassistant.const import (
-    ENERGY_KILO_WATT_HOUR,
-    PERCENTAGE,
-    POWER_WATT,
-    TEMP_CELSIUS,
-)
+from homeassistant.const import (ELECTRIC_CURRENT_AMPERE,
+                                 ELECTRIC_POTENTIAL_VOLT,
+                                 ENERGY_KILO_WATT_HOUR, PERCENTAGE, POWER_WATT,
+                                 TEMP_CELSIUS)
 from homeassistant.core import HomeAssistant
 from homeassistant.helpers.entity import EntityCategory
 from homeassistant.helpers.entity_platform import AddEntitiesCallback
 from homeassistant.helpers.typing import StateType
 from homeassistant.util.dt import utc_from_timestamp
+from pyfritzhome.fritzhomedevice import FritzhomeDevice
 
 from . import FritzBoxEntity
-from .const import CONF_COORDINATOR, DOMAIN as FRITZBOX_DOMAIN
+from .const import CONF_COORDINATOR
+from .const import DOMAIN as FRITZBOX_DOMAIN
```
> ... or at least make sure to follow our [style guidelines](https://developers.home-assistant.io/docs/development_guidelines)

with this I mean the [sorting of the imports](https://developers.home-assistant.io/docs/development_guidelines/#ordering-of-imports) and formatting of the code by using Black (_`black --fast homeassistant`_)

dreimer1986 replied:
Hopefully fixed now

dreimer1986 committed:
black --fast applied

mib1185 commented on homeassistant/components/fritzbox/sensor.py:
```diff
@@ -17,6 +15,8 @@
 )
 from homeassistant.config_entries import ConfigEntry
 from homeassistant.const import (
+    ELECTRIC_CURRENT_AMPERE,
+    ELECTRIC_POTENTIAL_VOLT,
```
every change in the import section -besides these to additions (_`ELECTRIC_CURRENT_AMPERE` and `ELECTRIC_POTENTIAL_VOLT`_)- should not be there

mib1185 replied:
```
root ➜ /workspaces/homeassistant-core (dreimer1986-patch-1) $ isort homeassistant/components/fritzbox/sensor.py 
Fixing /workspaces/homeassistant-core/homeassistant/components/fritzbox/sensor.py
```

mib1185 commented:
further please also add these two new sensors to the list in the [docs](https://www.home-assistant.io/integrations/fritzbox/#sensors)

mib1185 committed:
fix issort

mib1185 reviewed (approved):
as soon as a PR for the docs is available, it can be merged
thanks @dreimer1986 👍

mib1185 commented on homeassistant/components/fritzbox/sensor.py:
```diff
@@ -87,6 +89,26 @@ class FritzSensorEntityDescription(
+        native_value=lambda device: device.voltage / 1000 if device.voltage else 0.0,
+    ),
+    FritzSensorEntityDescription(
+        key="electric_current",
+        name="Electric Current",
```
the docs PR shows me, that "Voltage" is better name for this sensor

mib1185 committed:
rename electric potential to voltage
